### PR TITLE
Optimize halo exchange.

### PR DESCRIPTION
**Pull request by sriharshakandala**
- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: src/Spaces/dss2.jl [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+                val = perimeter_slab1[ip1] ⊞ perimeter_slab2[ip2]
+                perimeter_slab1[ip1] = val
+                perimeter_slab2[ip2] = val
+            end
+        end
+    end
+    return nothing
+end
+
+function dss_local_ghost2!(perimeter_data, perimeter, topology)
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:272]: does "local" part of ghost vertex dss (e.g. where 2 local elements share a vertex that needs to be computed). This is stored back in all locations in `perimeter_data`

File: src/Spaces/dss2.jl [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+        send_buf_idx,
+        recv_buf_idx,
+        scalaridx,
+        scalarfidx,
+        vectoridx,
+        vectorfidx,
+    )
+end
+
+function fill_send_buffer2!(dss_buffer::DSSBuffer)
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:420]: loads send_buffer from perimeter_data

File: src/Spaces/dss2.jl [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+end
+
+function dss_untransform2!(dss_buffer, data, local_geometry, perimeter)
+    (; vectorfidx, perimeter_data) = dss_buffer
+    ∂x∂ξ = local_geometry.∂x∂ξ
+    dss_untransform2!(perimeter_data, data, ∂x∂ξ, perimeter, vectorfidx)
+    return nothing
+end
+
+function dss_local2!(perimeter_data, perimeter, topology)
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:233]: local dss, written back to perimeter

File: src/Spaces/dss2.jl [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+        @inbounds ip = send_buf_idx[i, 2]
+        for f in 1:Nf, v in 1:Nv
+            @inbounds send_data[ctr] = pdata[v, ip, f, lidx]
+            ctr += 1
+        end
+    end
+    return nothing
+end
+
+function load_from_recv_buffer2!(dss_buffer::DSSBuffer)
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:438]: adds dss to representative element in perimeter

File: src/Spaces/dss2.jl [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+                    perimeter_slab = slab(perimeter_data, level, lidx)
+                    perimeter_slab[ip] = sum_data
+                end
+            end
+        end
+    end
+    return nothing
+end
+
+function dss_ghost2!(perimeter_data, perimeter, topology)
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:302]: takes representative element, and sets all colocated vertices to same value

File: src/Spaces/dss2.jl [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+                            pdata[level, ip, jp, fidx + 1, elem]
+                        ) * pweight[ip, jp, 1, elem]
+                end
+            end
+        end
+    end
+    return nothing
+end
+
+function dss_untransform2!(
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:137]: untransforms, and copies back

File: src/Topologies/dtopology2d.jl [hunk 7]
```diff
@@ -84,6 +84,30 @@ struct DistributedTopology2D{
...
+    "internal local elements (lidx)"
+    internal_elems::Vector{Int}
+    "local elements (lidx) located on process boundary"
+    perimeter_elems::Vector{Int}
+    "total number of unique non-boundary vertices"
+    nglobalvertices::Int
+    "total number of unique non-boundary faces"
+    nglobalfaces::Int
+    "global id for each unique ghost vertex"
+    ghost_vertex_gidx::Vector{Int}
```
Comment (simonbyrne) [inline src/Topologies/dtopology2d.jl:96]: Can we call this `gcidx` (connection) to distinguish from element `gidx`?

File: src/Topologies/dtopology2d.jl [hunk 8]
```diff
@@ -84,6 +84,30 @@ struct DistributedTopology2D{
...
+    "global id for each unique ghost face"
+    ghost_face_gidx::Vector{Int}
+    "number of unique vertices to be communicated to neighboring processes"
+    comm_vertex_lengths::Vector{Int}
```
Comment (simonbyrne) [inline src/Topologies/dtopology2d.jl:100]: `comm_vertex_lengths[i]` = number of unique vertices to be communicated to `neighbor_pids[i]`

File: src/Topologies/dtopology2d.jl [hunk 9]
```diff
@@ -84,6 +84,30 @@ struct DistributedTopology2D{
...
+    "number of unique faces to be communicated to each neighboring process"
+    comm_face_lengths::Vector{Int}
+    "neighbor process locations in neighbor_pids for each of the ghost vertices"
+    ghost_vertex_neighbor_loc::Vector{Int}
+    "offset array for `ghost_vertex_neighbor_loc`"
+    ghost_vertex_comm_idx_offset::Vector{Int}
```
Comment (simonbyrne) [inline src/Topologies/dtopology2d.jl:106]: a ragged array representation:
for `i = 1:length(ghost_vertex_gidx)`, we have that for `j = ghost_vertex_comm_idx_offset[i]:ghost_vertex_comm_idx_offset[i+1]-1`
the ghost vertex `ghost_vertex_gidx[i]` should get sent to `neighbor_pids[ghost_vertex_neighbor_loc[j]]`.

Comment (simonbyrne) [reply src/Topologies/dtopology2d.jl:108]: include this in the docstring

File: src/Spaces/dss.jl [hunk 10]
```diff
@@ -824,3 +824,4 @@ end
 function weighted_dss_ghost!(data, space, hspace, ghost_buffer)
     return data
 end
+include("dss2.jl")
```
Comment (simonbyrne) [inline src/Spaces/dss.jl:827]: move this include to `Spaces.jl`

File: src/Spaces/dss2.jl [hunk 11]
```diff
@@ -0,0 +1,624 @@
...
+        data::DataLayouts.IJFH,
+        ∂ξ∂x::DataLayouts.IJFH,
+        weight,
+        perimeter,
+        scalarfidx,
+        vectorfidx,
+        localelems,
+    )
+
+    Transforms vectors from Covariant axes to physical (local axis),
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:57]: don't indent this (it will render as code)

File: src/Spaces/dss2.jl [hunk 12]
```diff
@@ -0,0 +1,624 @@
...
+        ∂ξ∂x::DataLayouts.IJFH,
+        weight,
+        perimeter,
+        scalarfidx,
+        vectorfidx,
+        localelems,
+    )
+
+    Transforms vectors from Covariant axes to physical (local axis), 
+weights the data at perimeter nodes, and stores result in the perimeter_data array.
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:58]: explain what all the arguments are

File: src/Spaces/dss2.jl [hunk 13]
```diff
@@ -0,0 +1,624 @@
...
+)
+    pdata = parent(data)
+    pweight = parent(weight)
+    p∂ξ∂x = parent(∂ξ∂x)
+    pperimeter_data = parent(perimeter_data)
+    (i11, i12, i21, i22) =
+        length(p∂ξ∂x) / prod(size(∂ξ∂x)) == 4 ? (1, 3, 2, 4) : (1, 4, 2, 5)
+
+    @inbounds for elem in localelems
+        @inbounds for (p, (ip, jp)) in enumerate(perimeter)
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:78]: one `@inbounds is enough (it is applied syntactically to everything below)

File: src/Spaces/dss2.jl [hunk 14]
```diff
@@ -0,0 +1,624 @@
...
+            end
+
+        end
+    end
+    return nothing
+end
+"""
+    dss_transform2!(
+        perimeter_data,
+        data::DataLayouts.VIJFH,
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:107]: you only need one docstring for a function if all the methods do the same thing

File: src/Spaces/dss2.jl [hunk 15]
```diff
@@ -0,0 +1,624 @@
...
+    dss_untransform2!(
+        perimeter_data,
+        data::DataLayouts.IJFH,
+        ∂x∂ξ::DataLayouts.IJFH,
+        perimeter,
+        scalarfidx,
+        vectorfidx,
+    )
+
+Transforms the DSS`d local vectors back to Covariant12 vectors, and copies the DSS'd data from the
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:181]:
```suggestion
Transforms the DSS'd local vectors back to Covariant12 vectors, and copies the DSS'd data from the
```

File: src/Spaces/dss2.jl [hunk 16]
```diff
@@ -0,0 +1,624 @@
...
+    dss_untransform2!(
+        perimeter_data,
+        data::DataLayouts.VIJFH,
+        ∂x∂ξ::DataLayouts.VIJFH,
+        perimeter,
+        scalarfidx,
+        vectorfidx,
+    )
+
+Transforms the DSS`d local vectors back to Covariant12 vectors, and copies the DSS'd data from the
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:231]: again, this docstring is redundant

File: src/Spaces/dss2.jl [hunk 17]
```diff
@@ -0,0 +1,624 @@
+using DocStringExtensions
+
+"""
+    weighted_dss2!(data, space, dss_buffer)
+
+Computes weighted dss of `data`.
```
Comment (simonbyrne) [inline src/Spaces/dss2.jl:6]: Can you describe a bit more how it works, and link to the subfunctions?

Comment (simonbyrne) [reply src/Spaces/dss2.jl:6]: Also, need to add all the docstrings to `docs/src/api.md`

**Review by simonbyrne — CHANGES_REQUESTED** [review]
Can you look at the docs, and fix the formatting and duplicated docstrings, e.g.
https://clima.github.io/ClimaCore.jl/previews/PR965/api/#ClimaCore.Spaces.dss_transform2!

**Comment by sriharshakandala** [comment]
bors try

File: src/DataLayouts/DataLayouts.jl [hunk 18]
```diff
@@ -1253,6 +1253,21 @@ end
     dataview = @inbounds view(parent(data), v, Ni * z2 .+ (1:Ni))
     return dataview
 end
+# An iterator for datalayouts
+function Base.iterate(data::DataLayouts.AbstractData{S}, ivar = 1) where {S}
```
Comment (simonbyrne) [inline src/DataLayouts/DataLayouts.jl:1257]: If this isn't being used, perhaps remove it? I'm not sure that makes the most sense as a way to iterate.

File: src/Fields/Fields.jl [hunk 19]
```diff
@@ -378,8 +394,21 @@ function Spaces.create_ghost_buffer(field::Field)
         space.horizontal_space : space
     Spaces.create_ghost_buffer(field_values(field), hspace.topology)
 end
+#----------test code
```
Comment (simonbyrne) [inline src/Fields/Fields.jl:397]: is this comment still relevant?

**Review by simonbyrne — APPROVED** [review]
LGTM. Squash the commits before merging.

**Comment by sriharshakandala** [comment]
bors try

**Pull request merged**